Add Timeline plugin

hansthen opened the pull request:
I closed the earlier pull request for the `timeline` functionality as I was not happy with the approach. I think a separate plugin is better. I think the difference in functionality is sufficient to warrant its own plugin.

Besides the ability to use `end` time, this plugin also does not require a array of `times` array with matching elements in a `geometry` array. Just giving each feature a `start` and `end` is simpler.

Another `timeline` specific functionality, is that it allows multiple layers to be controlled by a single `TimeLineSlider` control.

Lastly, the `Timeline` class inherits from `GeoJson`, so it can use `GeoJsonPopup` and `GeoJsonTooltip` objects.

hansthen commented:
Btw, if you think this plugin is best served in its own repository, I am fine with that as well.

Conengmo commented:
Sorry for the delay in reviewing, I will get to it shortly.

hansthen commented:
pre-commit.ci autofix

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+    --------
+
+    >>> data = requests.get(
+    ...     "https://raw.githubusercontent.com/python-visualization/folium-example-data/main/borders.json"
```
this data doesn't exist

hansthen replied:
Done

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+    ...     data,
+    ...     style=JsCode(
+    ...         '''
+    ...        function (data) {
+    ...            function getColorFor(str) {
+    ...                // java String#hashCode
+    ...                var hash = 0;
+    ...                for (var i = 0; i < str.length; i++) {
+    ...                    hash = str.charCodeAt(i) + ((hash << 5) - hash);
+    ...                }
+    ...                var red = (hash >> 24) & 0xff;
+    ...                var grn = (hash >> 16) & 0xff;
+    ...                var blu = (hash >> 8) & 0xff;
+    ...                return "rgb(" + red + "," + grn + "," + blu + ")";
+    ...            }
+    ...            return {
+    ...                stroke: false,
+    ...                color: getColorFor(data.properties.name),
+    ...                fillOpacity: 0.5,
+    ...            };
```
is this necessary for the example? The example should be the minimum necessary to show how the class works. Or show case some common features. This elaborate custom styling doesn't seem that relevant for learning how the arguments work.

hansthen replied:
It was in the original example for the timeline plugin. I tried the example without colors, but the demo looked horrible.

hansthen replied:
Changed for a simpler example.

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+        """
+        {% macro header(this,kwargs) %}
+            <style>
+                .leaflet-bottom.leaflet-left {
+                    width: 100%;
+                }
+                .leaflet-control-container .leaflet-timeline-controls {
+                    box-sizing: border-box;
+                    width: 100%;
+                    margin: 0;
+                    margin-bottom: 15px;
+                }
```
some of this styling seems general to Leaflet. It may impact other features, so I'd rather not have it here. Better to make it apply to this plugin only, or include it in the map styling. But for the latter that's possibly dangerous, since it may break other features.

hansthen replied:
I do not know how to apply css only to a single plugin.

I am also not sure how much of a problem it is. I think Folium is typically run inside a separate iframe, so general styling will be limited to a single instance. I ran this `timeline` together with the `timestamped_geojson` in a single Jupyter notebook and there was no interference. 

The second css rule is limited to the "leaflet-timeline-controls" class, which can only be a problem if someone decides to reuse that class.

For the first css rule: the `TimelineSlider` will take the bottom left controller position, so no other controls can be positioned there for this map. To make the slider take the full width, which looks much nicer, this styling is required. This will impact other controls that take the bottom left, but I think in practice it will not be a problem.

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+    def render(self, **kwargs):
+        assert isinstance(
+            self._parent, Map
+        ), "Timeline can only be added to a Map object."
```
is that true? So it can't be grouped in a featuregroup?

hansthen replied:
Changed.

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+        show_ticks: bool = True,
+        steps: int = 1000,
+        playback_duration: int = 10000,
+        speed_slider: bool = True,
```
please make sure you include all arguments in the docstring.

hansthen replied:
Removed speed_slider as it was not used.

Conengmo reviewed (changes requested):
I'd like to have this plugin in our repo, since it seems like a better approach than our current timestamped geojson plugin. But some work is needed to get it mergeable.

I added some comments based on the code alone, I didn't test it yet because the example is missing data.

We'll also need:
- an entry in our documentation
- list this plugin in the 'plugins' page in the docs. We'll have to make sure that the description of this one and of timestamped geojson show the difference between the two.
- importing both classes in folium/plugins/__init__.py

hansthen commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
 branca>=0.6.0
+isodate
```
Done

hansthen commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+    default_js = [
+        (
+            "timeline",
+            "https://skeate.dev/Leaflet.timeline/examples/leaflet.timeline.js",
```
Done

hansthen commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,313 @@
+
+    def __init__(
+        self,
+        data: Union[dict, str, TextIO],
+        date_options: str = "YYYY-MM-DD HH:mm:ss",
+        point_to_layer: Optional[JsCode] = None,
+        style: Optional[JsCode] = None,
+        on_each_feature: Optional[JsCode] = None,
+        get_interval: Optional[JsCode] = None,
```
Done

hansthen commented on docs/user_guide/plugins/timeline.md:
This is temporary so the tests pass. Once the PR is accepted I will change this to point to data inside our own repository.

hansthen commented:
I added documentation and tests. I also updated the documentation of timestamped_geojson to make the reader aware of the two plugins. The tests (for now) refer to an example data file (borders.json) in the `skeate` repo. This is just to make the tests pass until this PR is accepted. I will update it to point to the Folium repo once the PR is accepted.

hansthen commented:
Any chance to have a look at the updated PR?

hansthen commented:
@Conengmo @ocefpaf I know Folium is a volunteer project and I greatly appreciate your work on it. It is hard to make time free next to work and family obligations. Thank you both for the effort you put in the project!

I feel a bit guilty for adding to your workload by spamming you with Pull Requests. Is there any way I can help that will lessen the demands placed on you?

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,289 @@
+      data: file, dict or str.
+          The geojson data you want to plot.
+
+      get_interval: JsCode
```
```suggestion
      get_interval: JsCode, optional
```

Conengmo replied:
You marked this as resolved, but looks like the docstring didn't change? `get_interval` is an optional parameter isn't it?

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,289 @@
+        self,
+        # arguments relevant to both interval and timestamp mode
+        auto_play: bool = True,
+        date_options: str = "YYYY-MM-DD HH:mm:ss",
```
this argument is not in the docstring

hansthen replied:
Done.

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,289 @@
+        ), "TimelineSlider can only be added to a Map object."
+        super().render(**kwargs)
+
+    def _get_self_bounds(self):
+        """
+        Computes the bounds of the object itself (not including it's children)
+        in the form [[lat_min, lon_min], [lat_max, lon_max]].
+
+        """
+        return get_bounds(self.data, lonlat=True)
```
this method uses `self.data`, but that attribute doesn't exist. Maybe remove this method?

hansthen replied:
Oops.

Conengmo commented on folium/plugins/timeline.py:
```diff
@@ -0,0 +1,289 @@
+    ):
+        super().__init__()
+        self._name = "TimelineSlider"
+        self._getDisplayDateFormat = JsCode(
```
why is this an instance attribute on `self`? Seems like it is only used for `kwargs["format_output"]`. In that case I'd put it directly there, not storing it in `self` as well.

hansthen replied:
Done.

Conengmo reviewed (changes requested):
Thanks for addressing my comments! Documentation looks good 👍  I think we're really close here, some minor comments and proposed steps for hosting the example data ourselves first. After that we should be able to merge this one!

Conengmo commented:
> I know Folium is a volunteer project and I greatly appreciate your work on it. It is hard to make time free next to work and family obligations. Thank you both for the effort you put in the project!

Thanks for your kind words, I really appreciate it 🙏 

> I feel a bit guilty for adding to your workload by spamming you with Pull Requests. Is there any way I can help that will lessen the demands placed on you?

Appreciate the question, but I'm happy to have high quality contributions, that's not spam! I have two thoughts. First to make PRs that are easy to review and merge. Clear description, complete, and most importantly simple and concise, so no unnecessary stuff and simple code. Second to help keep Folium maintainable in the future by being critical of the tradeoff between added complexity and new features.

If you're interested, you can also become more involved with maintaining Folium. Helping triage issues would be very useful. And also help review PRs other than your own. Are you interested in something like that?

hansthen commented:
> If you're interested, you can also become more involved with maintaining Folium. Helping triage issues would be very useful. And also help review PRs other than your own. Are you interested in something like that?

I can help with both. I did start with the issue list to see if there were things I could work on or reproduce. For code reviews, how would you prefer to start? Will you use the github "request code review" feature to assign PRs to me?

Conengmo commented:
> I can help with both.

That's great! I added you as a collaborator on Folium. You can now also close issues or apply labels. And you can merge PRs as well. Please use this responsibly: not merging your own code without review, and for now to start get a second opinion before merging anything consequential.

> For code reviews, how would you prefer to start? Will you use the github "request code review" feature to assign PRs to me?

Yes, but you can also take initiative to review any issue or PR that has your interest and that you feel you could help with.

Conengmo commented:
@ocefpaf @martinfleis wanted to let you know as well I invited @hansthen to become a collaborator, see https://github.com/python-visualization/folium/settings/access.

Conengmo reviewed (changes requested):
Looks like `examples/data/borders.json` is still in this PR.

Nearly there!

hansthen commented:
> Looks like `examples/data/borders.json` is still in this PR.
> 
> Nearly there!

I used the smaller copy from the examples (also updated in the tests).

Conengmo reviewed (approved):
Great, thanks!
